from pathlib import Path
import glob


gtf_latest_version_f=config['gtf_latest_version_f']
gtf_f=config['gtf_f']
ref_genome=config['ref_genome']

BAM_COND_MASK=[]
ALL_BAMS=[]

for cond in config['conds']:

    cond_bams=[Path(bam).stem.split('.')[0] for bam in glob.glob(config['realigned_bam_dir']+cond+"/*.bam")]


    ALL_BAMS+=cond_bams
    for bam in cond_bams:
        BAM_COND_MASK.append(cond)







rule all:
    input:
        expand(config['junc_dir']+"{cond}/{sample}.annot",zip,cond=BAM_COND_MASK,sample=ALL_BAMS),expand(config['junc_dir']+"{cond}/{sample}_v45.annot",zip,cond=BAM_COND_MASK,sample=ALL_BAMS)

rule bam2junc:
    input:
        config['realigned_bam_dir']+'{cond}/{sample}.Aligned.sortedByCoord.waspFiltered.out.bam'
    output:
        config['junc_dir']+"{cond}/{sample}.junc"
    params:
        min_intron_length=50,
        anchor_length=8,
        max_intron_length=500000

    resources:
        mem_mb=3000
    shell:
        "regtools junctions extract -s 0 -a {params.anchor_length} -m {params.min_intron_length} -M {params.max_intron_length} {input} -o {output}"

rule junc_annot:
    input:
        config['junc_dir']+"{cond}/{sample}.junc"
    output:
        config['junc_dir']+"{cond}/{sample}.annot"
    params:
        gtf_f=gtf_f,
        ref_genome=ref_genome

    resources:
        mem_mb=3000
    shell:
        "regtools junctions annotate {input} {params.ref_genome} {params.gtf_f} -o {output}"
